### PR TITLE
Improving docs structure

### PR DESCRIPTION
**Pull request by jrudolph**
I started restructuring the docs a bit from the rather flat structure to a bit more guided form.

It's still WIP.

### CODE REVIEW

**Comment by jrudolph** [comment]
> Quite a lot of 'gRPC' in titles now, that might be over-doing it?

Yeah, good point ;)

> I think we need a good "when and why to use gRPC" in the introduction. The comparing/contrasting with other technologies can be useful for that, but this is now moved to the "gRPC vs. The Rest" section at the bottom. That's fine, but then we need a new way to explain when you'd use gRPC for the introduction.

Yes, makes sense. In the existing page, the placing of the comparison seemed a bit arbitrary, so I moved it out.

File: docs/src/main/paradox/client.md [hunk 1]
```diff
@@ -1,302 +0,0 @@
-# Client
```
Comment (jrudolph) [inline docs/src/main/paradox/client.md]: Split up into several pages.

File: docs/src/main/paradox/server/index.md [hunk 2]
```diff
@@ -0,0 +1,9 @@
+# Providing Services (Server)
+
+@@ toc
+
+@@@ index
+
+ * [Walkthrough](walkthrough.md)
```
Comment (jrudolph) [inline docs/src/main/paradox/server/index.md]: That's still a bit empty right now. I think we might want to provide more pages with reference information that isn't a walkthrough.

File: docs/src/main/paradox/whygrpc.md [hunk 3]
```diff
@@ -0,0 +1,47 @@
+# Why gRPC?
```
Comment (jrudolph) [inline docs/src/main/paradox/whygrpc.md]: Added page "Why gRPC?" that contains the previous comparison + some extra motivation for gRPC.

File: docs/src/main/paradox/whygrpc.md [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+* Where SOAP is more flexible about transport, gRPC standardizes on HTTP/2.
+* Where in SOAP protocols are often set in stone once defined (often requiring a new path for every version of the service), Protobuf is explicitly intended to support schema evolution.
+
+## gRPC vs Message Bus
+
+* While built on an efficient non-blocking implementation, gRPC is still 'synchronous' in the sense that it requires both 'sides' of the communication to be available at the same time. When using a (persistent) message bus only the producer and the bus must be up, the consumer does not need to be available, leading to a higher degree of decoupling.
+* While gRPC supports bidirectional streaming for each request, when using a message bus the streams are decoupled
+
+## gRPC vs Akka Remoting
```
Comment (jrudolph) [inline docs/src/main/paradox/whygrpc.md]: I thought adding this section might make sense as we want to promote gRPC over Akka Remoting in the future.

**Comment by jrudolph** [comment]
Complete new structure:

screenshot from 2018-08-14 14-40-14

**Comment by jrudolph** [comment]
Rebased and ready to review.

File: docs/src/main/paradox/buildtools/index.md [hunk 5]
```diff
@@ -0,0 +1,11 @@
+# Build Tool Support
+
+@@toc
+
+@@@index
+ 
+ * [SBT](sbt.md)
```
Comment (ktoso) [inline docs/src/main/paradox/buildtools/index.md]: `Sbt` or `sbt`, we stopped writing it as "SBT" years ago AFAIR -- cc @eed3si9n

Comment (eed3si9n) [reply docs/src/main/paradox/buildtools/index.md]: I always write it as sbt, all lower-cased.

File: docs/src/main/paradox/client/walkthrough.md [hunk 6]
```diff
@@ -6,7 +6,7 @@ To get started, you must obtain the @ref[`.proto`](../proto.md) file(s) that des
 
 Add `.proto` files to your project's @sbt[`src/main/protobuf`]@gradle[`src/main/proto`]@maven[`src/main/proto`] directory.
 See the detailed chapters on @ref[sbt](../buildtools/sbt.md), @ref[Gradle](../buildtools/gradle.md) and @ref[Maven](../buildtools/maven.md) for information on picking up
- .proto definitions from dependencies automatically.
+`.proto` definitions from dependencies automatically.
```
Comment (ktoso) [inline docs/src/main/paradox/client/walkthrough.md]: *.proto I guess

File: plugin-tester-scala/src/main/scala/example/myapp/helloworld/GreeterClient.scala [hunk 7]
```diff
@@ -22,24 +22,27 @@ import scala.util.{ Failure, Success }
 object GreeterClient {
 
   def main(args: Array[String]): Unit = {
-
+    // Boot akka
```
Comment (chbatey) [inline plugin-tester-scala/src/main/scala/example/myapp/helloworld/GreeterClient.scala]: We need Akka-Boot :tm:

File: docs/src/main/paradox/buildtools/index.md (hunk 5 above)
Comment (jrudolph) [reply docs/src/main/paradox/buildtools/index.md]: I will fix it but it is an index label which isn't used anyway ;)

File: plugin-tester-scala/src/main/scala/example/myapp/helloworld/GreeterClient.scala (hunk 7 above)
Comment (jrudolph) [reply plugin-tester-scala/src/main/scala/example/myapp/helloworld/GreeterClient.scala]: :boot: Akka Boots :boot:

Comment (ktoso) [reply plugin-tester-scala/src/main/scala/example/myapp/helloworld/GreeterClient.scala]: https://www.youtube.com/watch?v=zuDtACzKGRs ?

File: docs/src/main/paradox/overview.md [hunk 8]
```diff
@@ -0,0 +1,38 @@
+# Overview
+
+## gRPC
+
+[gRPC](https://grpc.io) is a transport mechanism for request/response and (non-persistent) streaming use cases. See
+@ref[Why gRPC?](whygrpc.md) for more information about why to choose gRPC as your RPC transport.
```
Comment (raboof) [inline docs/src/main/paradox/overview.md]: Maybe 'about when to use gRPC as your transport', but that's nitpicking.

File: docs/src/main/paradox/whygrpc.md [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+It is a schema-first RPC framework, where your protocol is declared in a
+@ref[protobuf service descriptor](proto.md), and requests and responses will be streamed over an HTTP/2
+connection.
+
+It has several advantages:
+
+ * Schema-first design favors well-defined and decoupled service interfaces over brittle ad-hoc solutions.
+ * Protobuf-based wire protocol is efficient, well-known, and allows compatible schema evolution.
+ * Based on HTTP/2 which allows multiplexing several data streams over a single connection.
+ * Streaming requests / responses are first class.
```
Comment (raboof) [inline docs/src/main/paradox/whygrpc.md]: 'and'?

File: plugin-tester-scala/src/main/scala/example/myapp/helloworld/GreeterClient.scala (hunk 7 above)
Comment (jrudolph) [reply plugin-tester-scala/src/main/scala/example/myapp/helloworld/GreeterClient.scala]: :beer: :boot:

File: docs/src/main/paradox/whygrpc.md (hunk 4 above)
Comment (raboof) [reply docs/src/main/paradox/whygrpc.md]: Very! Perhaps also mention that remoting is more sensible to Akka versions and configuration, so using gRPC makes the applications more loosely coupled?

File: docs/src/main/paradox/whygrpc.md [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+
+## gRPC vs Message Bus
+
+* While built on an efficient non-blocking implementation, gRPC is still 'synchronous' in the sense that it requires both 'sides' of the communication to be available at the same time. When using a (persistent) message bus only the producer and the bus must be up, the consumer does not need to be available, leading to a higher degree of decoupling.
+* While gRPC supports bidirectional streaming for each request, when using a message bus the streams are decoupled
+
+## gRPC vs Akka Remoting
+
+* While Akka Remoting allows exchanging messages between Akka ActorSystems transparently, it still requires significant effort to support efficient and compatible message serialization.
+  Large messages can clog the message transport. In contrast to gRPC, streaming is not first-class but needs to be built on top of message passing (e.g. by using [StreamRefs](https://doc.akka.io/docs/akka/2.5.14/stream/stream-refs.html)).
```
Comment (raboof) [inline docs/src/main/paradox/whygrpc.md]: Perhaps link to 'current'? If we ever remove `stream-refs.html` we'll want to add redirects anyway.

File: plugin-tester-java/src/main/java/example/myapp/helloworld/GreeterServer.java [hunk 11]
```diff
@@ -39,20 +39,33 @@ public static void main(String[] args) throws Exception {
...
       Materializer mat = ActorMaterializer.create(sys);
 
+      // Instantiate implementation
       GreeterService impl = new GreeterServiceImpl(mat);
 
+      // Bind implementation to localhost:8080
       Http.get(sys).bindAndHandleAsync(
           GreeterServiceHandlerFactory.create(impl, mat),
-          ConnectWithHttps.toHostHttps("127.0.0.1", 8080).withCustomHttpsContext(serverHttpContext()),
+          // HTTP/2 servers are required to use TLS
```
Comment (raboof) [inline plugin-tester-java/src/main/java/example/myapp/helloworld/GreeterServer.java]: Perhaps link to https://doc.akka.io/docs/akka-http/current/server-side/http2.html#use-bindandhandleasync-and-https?

File: docs/src/main/paradox/whygrpc.md (hunk 10 above)
Comment (jrudolph) [reply docs/src/main/paradox/whygrpc.md]: Thanks, good catch.

File: plugin-tester-java/src/main/java/example/myapp/helloworld/GreeterServer.java (hunk 11 above)
Comment (jrudolph) [reply plugin-tester-java/src/main/java/example/myapp/helloworld/GreeterServer.java]: Maybe that can be done in the context of removing the TLS code, @chbatey?

File: docs/src/main/paradox/overview.md (hunk 8 above)
Comment (jrudolph) [reply docs/src/main/paradox/overview.md]: Thanks, done.

File: docs/src/main/paradox/whygrpc.md (hunk 9 above)
Comment (jrudolph) [reply docs/src/main/paradox/whygrpc.md]: Done.

**Comment by jrudolph** [comment]
Rebased...

File: docs/src/main/paradox/server.md [hunk 12]
```diff
@@ -1,220 +0,0 @@
...
-
-
-## Stateful services
-
-More often than not, the whole point of the implementing a service is to keep state. Since the service implementation
-is shared between concurrent incoming requests any state must be thread safe.
-
-There are two recommended ways to deal with this:
-
- * Make the state immutable – immutable state that is created before or when the service is instantiated and then never changes is safe
```
Comment (jrudolph) [inline docs/src/main/paradox/server.md]: I removed that one, btw. Maybe I'm missing something but "stateful" to me implies that the state can change over time?

Instead, I pointed to CRUD-style programming where a DB handles concurrency.

**Pull request merged**